File: hosts/android/feature/products/impl/src/main/java/io/paritytech/polkadotapp/feature_products_impl/domain/truapi/PocketDeeplinkParser.kt
package io.paritytech.polkadotapp.feature_products_impl.domain.truapi

import uniffi.truapi_server.NavigateDecision
import uniffi.truapi_server.parseNavigate
import javax.inject.Inject
import uniffi.truapi_server.PocketDeeplinkAction as NativePocketDeeplinkAction

enum class PocketDeeplinkAction { ADD, OPEN }

/** A `/-/pocket/<action>?card=<id>` deeplink as the core classifies it; [productHost] is the lower-cased dotNS name. */
data class PocketDeeplink(
    val productHost: String,
    val action: PocketDeeplinkAction,
    val cardId: String,
)

/**
 * Classifies through the core's `parse_navigate`, so every host reads a Pocket deeplink the same
 * way. Kept behind this adapter, like the host bridges, so a bindgen rename does not ripple.
 */
class PocketDeeplinkParser @Inject constructor() {
    fun parse(url: String): PocketDeeplink? {
        val decision = parseNavigate(url) as? NavigateDecision.Pocket ?: return null
        return PocketDeeplink(
            productHost = decision.identifier,
            action = when (decision.action) {
                NativePocketDeeplinkAction.ADD -> PocketDeeplinkAction.ADD
                NativePocketDeeplinkAction.OPEN -> PocketDeeplinkAction.OPEN
            },
            cardId = decision.cardId,
        )
    }
}

File: hosts/ios/polkadot-app/Modules/Products/Pocket/PocketDeeplinkParser.swift
import Foundation
import TrUAPIHost

/// Named to avoid the core's own `PocketDeeplinkAction`: a file importing both
/// cannot tell them apart, and the collision crashes the type checker rather
/// than producing a diagnostic.
enum PocketLinkAction: Equatable {
    case add
    case open
}

/// A `/-/pocket/<action>?card=<id>` link as the core classifies it.
/// `productHost` is the lower-cased dotNS name.
struct PocketDeeplink: Equatable {
    let productHost: String
    let action: PocketLinkAction
    let cardId: String
    /// The normalized `polkadot://` form, with the card id percent-encoded, so
    /// re-parsing it names the same card.
    let canonicalUrl: String
}

/// Classifies through the core's `parse_navigate`, so every host reads a Pocket
/// link the same way. Kept behind this adapter, like the host bridges, so a
/// bindgen rename does not ripple through the app.
struct PocketDeeplinkParser {
    func parse(_ url: String) -> PocketDeeplink? {
        guard case let .pocket(identifier, action, cardId, canonicalUrl) = parseNavigate(input: url) else {
            return nil
        }

        let linkAction: PocketLinkAction =
            switch action {
            case .add: .add
            case .open: .open
            }

        return PocketDeeplink(
            productHost: identifier,
            action: linkAction,
            cardId: cardId,
            canonicalUrl: canonicalUrl
        )
    }
}
